src/checkpoint/checkpoint.ts

```diff
@@ -104,16 +151,46 @@ export default class Checkpoint {
       await Promise.delay(12e3);
       return this.next(blockNum);
     }
+
     await this.handleBlock(block);
-    const query = 'UPDATE checkpoint SET number = ?';
-    await this.mysql.queryAsync(query, [block.block_number]);
+
+    await this.store.setMetadata('last_indexed_block', block.block_number);
+
     return this.next(blockNum + 1);
   }
 
-  private async handleBlock(block) {
+  private async getNextCheckpointBlock(blockNum: number): Promise<number | null> {
+    if (this.cpBlocksCache === null) {
+      // cache is null when we can no more find a record in the database
+      // so exiting early here to avoid polling the database in subsequent
+      // loops.
+      return null;
+    }
+
+    if (this.cpBlocksCache.length !== 0) {
+      return this.cpBlocksCache.shift();
+    }
+
+    const checkpointBlocks = await this.store.getNextCheckpointBlocks(
+      blockNum,
+      this.sourceContracts
+    );
+    if (checkpointBlocks.length === 0) {
+      this.log.info({ blockNumber: blockNum }, 'no more checkpoint blocks in store');
+      // disabling cache to stop polling database
+      this.cpBlocksCache = null;
+
+      return null;
+    }
+
+    this.cpBlocksCache = checkpointBlocks;
+    return this.cpBlocksCache.shift();
+  }
+
+  private async handleBlock(block: GetBlockResponse) {
     this.log.info({ blockNumber: block.block_number }, 'handling block');
 
-    for (const receipt of block.transaction_receipts) {
+    for (const receipt of Object.values(block.transaction_receipts)) {
```

Comment:
Any reason to just return values here? Wouldn't this break the writer functions?

Reply:
I actually forgot to revert this, I planned to address it later, but due to me reference the type for `block`, typescript was complaining because the type definition for [transaction_receipts](https://github.com/0xs34n/starknet.js/blob/develop/src/types/api.ts#L156-L168) doesn't indicate that the field implements an Iterable Iterator. So Object.values() was an easy fix. And it doesn't break the writer function.

But there is no need for the unnecssary Object.values() call, so my plan was to ts-ignore it for now.

Reply:
Fixed here